cpp/src/reductions/reductions.cu

```diff
@@ -187,80 +197,54 @@ struct DeviceMax {
     T operator() (const T &lhs, const T &rhs) {
         return lhs >= rhs? lhs: rhs;
     }
-};
 
-#define DEF_REDUCE_OP_NUM(F)                                                      \
-gdf_error F##_generic(gdf_column *col, void *dev_result,                          \
-                          gdf_size_type dev_result_size) {                        \
-    switch ( col->dtype ) {                                                       \
-    case GDF_FLOAT64: return F##_f64(col, (double*)dev_result, dev_result_size);  \
-    case GDF_FLOAT32: return F##_f32(col, (float*)dev_result, dev_result_size);   \
-    case GDF_INT64:   return F##_i64(col, (int64_t*)dev_result, dev_result_size); \
-    case GDF_INT32:   return F##_i32(col, (int32_t*)dev_result, dev_result_size); \
-    case GDF_INT8:    return F##_i8(col,  (int8_t*)dev_result, dev_result_size);  \
-    default:          return GDF_UNSUPPORTED_DTYPE;                               \
-    }                                                                             \
-}
+    template<typename T>
+    static constexpr T identity = std::numeric_limits<T>::lowest();
+};
 
-#define DEF_REDUCE_OP_REAL(F)                                                     \
-gdf_error F##_generic(gdf_column *col, void *dev_result,                          \
-                          gdf_size_type dev_result_size) {                        \
-    switch ( col->dtype ) {                                                       \
-    case GDF_FLOAT64: return F##_f64(col, (double*)dev_result, dev_result_size);  \
-    case GDF_FLOAT32: return F##_f32(col, (float*)dev_result, dev_result_size);   \
-    default:          return GDF_UNSUPPORTED_DTYPE;                               \
-    }                                                                             \
-}
+template <typename Op>
+struct ReduceDispatcher {
+    template <typename T,
+              typename std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
+    gdf_error operator()(gdf_column *col, void *dev_result, gdf_size_type dev_result_size) {
+        T identity = Op::template identity<T>;
```

Comment:
This line is failing to compile with gcc5.4. Minimal repro here: https://wandbox.org/permlink/4fv7vIJpvyTjw6eK

I did some digging, and discovered that it's a compiler bug in the GCC5 family: https://baptiste-wicht.com/posts/2017/08/simplify-your-type-traits-with-c%2B%2B14-variable-templates.html

We will either need to find a different solution, or move to gcc6+ (there was an effort to move to gcc7.2 that got backburnered)

Reply:
Thanks for digging. Submitted a change to function template instead of variable template. CI running.